server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -245,10 +251,12 @@ public UnassignedInfo(Reason reason, @Nullable String message, @Nullable Excepti
         this.failure = failure;
         this.failedAllocations = failedAllocations;
         this.lastAllocationStatus = Objects.requireNonNull(lastAllocationStatus);
+        this.failedNodeIds = Collections.unmodifiableSet(failedNodeIds);
         assert (failedAllocations > 0) == (reason == Reason.ALLOCATION_FAILED) :
             "failedAllocations: " + failedAllocations + " for reason " + reason;
         assert !(message == null && failure != null) : "provide a message if a failure exception is provided";
         assert !(delayed && reason != Reason.NODE_LEFT) : "shard can only be delayed if it is unassigned due to a node leaving";
+        assert failedAllocations >= failedNodeIds.size() : "failedAllocations: " + failedAllocations + " failedNodeIds: " + failedNodeIds;
```

Comment:
Unfortunately this does not hold when we start carrying over the failedNodeIds in `processExistingRecoveries` (unless we also carry over the failedAllocations, which is reset there).

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/48265/commits/89c675346b2d932bc07d4f94932766e52c57e2e6.